Improve docs (#310)
trixi-framework/Trixi.jl: docs/src/performance.md

# Performance

Trixi.jl is designed to balance performance and readability. Since Julia provides

> a lot of zero-cost abstractions, it's often possible to optimize both goals

Reviewer: ```suggestion
a lot of zero-cost abstractions, it is often possible to optimize both goals
```

simultaneously.

> A usual development workflow in Julia is

Reviewer: ```suggestion
The usual development workflow in Julia is
```

1. Make it work.
2. Make it fast.

To achieve the second step, you should be familiar with (at least) the section on
[performance tips in the Julia manual](https://docs.julialang.org/en/v1/manual/performance-tips/).
Here, we just list some important aspects you should consider when developing Trixi.

- Consider using `@views`/`view(...)` when using array slices, except on the left-side
  of an assignment
  ([further details](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-views)).

> - Functions are essentially for free, since they can usually be inlined using `@inline`
>   ([further details](https://docs.julialang.org/en/v1/manual/performance-tips/#Break-functions-into-multiple-definitions)).

Reviewer: ```suggestion
- Functions are essentially for free, since they are usually automatically inlined where it makes sense (using `@inline` can be used as an additional hint to the compiler)
  ([further details](https://docs.julialang.org/en/v1/manual/performance-tips/#Break-functions-into-multiple-definitions)).
```
I'd like to make it explicit that `@inline` does not require the compiler to inline, nor is it required to inline a function. Working with students for many years with C++ and seeing their misconceptions about `inline`, I feel like this has to be emphasized (in C++ it is even worse since there `inline` has a linkage-related meaning).
